packages/library-base/src/commonMain/kotlin/io/realm/internal/query/SingleQuery.kt

```diff
@@ -30,11 +36,32 @@ internal class SingleQuery<E : RealmObject> constructor(
         return model as E
     }
 
-    override fun asFlow(): Flow<E?> {
+    override fun asFlow(): Flow<ObjectChange<E>> {
         realmReference.checkClosed()
+
+        var head: E? = null
+        var headFlow: Cancellable? = null
+
         return realmReference.owner.registerObserver(this)
-            .map { results ->
-                results.list.firstOrNull()
+            .filter { listChange: ListChange<RealmResultsImpl<E>> ->
+                // Denotes when to subscribe to the next object:
+                // A change on the head of the list, and it is not the same as the previous head
+                val newHead: E? = listChange.list.firstOrNull()
+
+                (newHead != null && !newHead.hasSameObjectKey(head)).also {
+                    head = newHead
+                }
+            }.flatMapMerge { listChange ->
+                // Don't close the flow if the head was removed
+                if (listChange !is UpdatedList<*> || !listChange.deletions.contains(0))
```

Comment:
Given the logic above, we only get here if the event is a `UpdatedList` event, correct? So perhaps it would be cleaner to just turn the first if into an assert `val change = listChange as UpdatedList` (or something).

Since this is a query result, we should never see a `DeletedList` I think?

Reply:
This actually reminds me, that at some point I was considering if we should have separate type hierarchies for `List` and `Query` since `DeleteList` doesn't really make sense for queries 🤔